editor/src/components/inspector/controls/color-picker-swatches.tsx

```diff
@@ -0,0 +1,185 @@
+import * as React from 'react'
+import { Button, colorTheme, FlexColumn, FlexRow } from '../../../uuiui'
+import { v4 as UUID } from 'uuid'
+import { Substores, useEditorState } from '../../editor/store/store-hook'
+import { useDispatch } from '../../editor/store/dispatch-context'
+import { updateColorSwatches } from '../../editor/actions/action-creators'
+import { ColorSwatch } from '../../editor/store/editor-state'
+
+export interface ColorPickerSwatchesProps {
+  onSelectColor: (hex: string) => void
+  currentColor: string
+}
+
+const SWATCH_SIZE = 24
+
+export const ColorPickerSwatches = React.memo((props: ColorPickerSwatchesProps) => {
+  const { currentColor, onSelectColor } = props
+
+  const storeColorSwatches = useEditorState(
+    Substores.restOfEditor,
+    (store) => store.editor.colorSwatches,
+    'ColorPickerPresets color presets',
+  )
+
+  const [editing, setEditing] = React.useState(false)
+  const [colorSwatches, setColorSwatches] = React.useState<Array<ColorSwatch>>(storeColorSwatches)
+
+  const dispatch = useDispatch()
+
+  React.useEffect(() => {
+    dispatch([updateColorSwatches(colorSwatches)], 'everyone')
+  }, [colorSwatches, dispatch])
+
+  React.useEffect(() => {
+    if (colorSwatches.length <= 0) {
+      setEditing(false)
+    }
+  }, [colorSwatches])
+
+  const toggleEditing = React.useCallback(() => {
+    setEditing(!editing)
+  }, [editing])
+
+  const onAddPreset = React.useCallback(() => {
+    const preset: ColorSwatch = {
+      id: UUID(),
+      hex: currentColor,
+    }
+    setColorSwatches(colorSwatches.concat(preset))
+  }, [currentColor, colorSwatches])
+
+  const onClickPreset = React.useCallback(
+    (c: ColorSwatch) => () => {
+      if (editing) {
+        setColorSwatches(colorSwatches.filter((p) => p.id != c.id))
+      } else {
+        onSelectColor(c.hex)
+      }
+    },
+    [editing, onSelectColor, colorSwatches],
+  )
+
+  return (
+    <FlexColumn
+      style={{
+        padding: '0px 8px 4px 8px',
+        marginTop: 4,
+        gap: 4,
+      }}
+    >
+      <FlexRow style={{ justifyContent: 'space-between' }}>
+        <div style={{ fontWeight: 600 }}>Swatches</div>
+        {colorSwatches.length > 0 && (
+          <Button spotlight highlight style={{ padding: '0 6px' }} onMouseDown={toggleEditing}>
+            {editing ? 'Done' : 'Edit'}
+          </Button>
+        )}
+      </FlexRow>
+
+      <FlexRow style={{ flexWrap: 'wrap', gap: 2 }}>
+        {colorSwatches.map((c, index) => {
+          return (
+            <Button
+              onClick={onClickPreset(c)}
+              key={index}
```

Comment:
index as key should be last resort, it's the same as not providing a key at all.

Reply:
ah yeah good spot, it's a remnant of an early pass, fixing

Reply:
https://github.com/concrete-utopia/utopia/pull/3133/commits/22f1a93248cfe5a09cbb7f1b261e2974479a1111